src/lib/optimizer/strategy/predicate_split_up_rule.cpp

```diff
@@ -2,44 +2,90 @@
   Assert(root->type == LQPNodeType::Root, "PredicateSplitUpRule needs root to hold onto");
 
-  /**
-   * Step 1:
-   *    - Collect PredicateNodes that can be split up into multiple ones into `predicate_nodes_to_flat_conjunctions`
-   */
-  auto predicate_nodes_to_flat_conjunctions =
-      std::vector<std::pair<std::shared_ptr<PredicateNode>, std::vector<std::shared_ptr<AbstractExpression>>>>{};
-
+  auto predicate_nodes = std::vector<std::shared_ptr<PredicateNode>>{};
   visit_lqp(root, [&](const auto& sub_node) {
     if (const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(sub_node)) {
-      const auto flat_conjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::And);
-
-      if (flat_conjunction.size() > 1) {
-        predicate_nodes_to_flat_conjunctions.emplace_back(predicate_node, flat_conjunction);
-      }
+      predicate_nodes.emplace_back(predicate_node);
     }
-
     return LQPVisitation::VisitInputs;
   });
 
+  // _split_conjunction() and _split_disjunction() split up logical expressions by calling each other recursively
+  for (const auto& predicate_node : predicate_nodes) {
+    if (!_split_conjunction(predicate_node)) {
+      // If there is no conjunction at the top level, try to split disjunction first
+      _split_disjunction(predicate_node);
+    }
+  }
+}
+
+bool PredicateSplitUpRule::_split_conjunction(const std::shared_ptr<PredicateNode>& predicate_node) const {
+  const auto flat_conjunction = flatten_logical_expressions(predicate_node->predicate(), LogicalOperator::And);
+  if (flat_conjunction.size() <= 1) {
+    return false;
+  }
+
   /**
-   * Step 2:
-   *    - Split up qualifying PredicateNodes into multiple consecutive PredicateNodes. We have to do this in a
-   *      second pass, because manipulating the LQP within `visit_lqp()`, while theoretically possible, is prone to
-   *      bugs.
+   * Split up PredicateNode with conjunctive chain (e.g., `PredicateNode(a AND b AND c)`) as its scan expression into
+   * multiple consecutive PredicateNodes (e.g. `PredicateNode(c) -> PredicateNode(b) -> PredicateNode(a)`).
    */
-  for (const auto& [predicate_node, flat_conjunction] : predicate_nodes_to_flat_conjunctions) {
-    for (const auto& predicate_expression : flat_conjunction) {
-      lqp_insert_node(predicate_node, LQPInputSide::Left, PredicateNode::make(predicate_expression));
-    }
-    lqp_remove_node(predicate_node);
+  for (const auto& predicate_expression : flat_conjunction) {
+    const auto& new_predicate_node = PredicateNode::make(predicate_expression);
+    lqp_insert_node(predicate_node, LQPInputSide::Left, new_predicate_node);
+    _split_disjunction(new_predicate_node);
+  }
+  lqp_remove_node(predicate_node);
+
+  return true;
+}
+
+void PredicateSplitUpRule::_split_disjunction(const std::shared_ptr<PredicateNode>& predicate_node) const {
+  if (!_should_split_disjunction) {
```

Comment:
This is not the right place for this check. It is not the job of `_split_disjunction` to decide whether it should do something or not - that is the job of the caller.